pkgs/stdenv/generic/setup.sh

```diff
@@ -47,12 +47,21 @@ getAllOutputNames() {
     fi
 }
 
-# Logs arguments to $NIX_LOG_FD, if it exists, no-op if it does not.
+# Accepts two or more arguments. The first argument is the log level (see `NIX_DEBUG`).
+# The rest of the arguments are logged to $NIX_LOG_FD, if it exists.
+# The function does nothing if it $NIX_LOG_FD is not set.
+# TODO: use structured logging (`@nix { "action": "msg", "level": $level, "msg": $@ }`)
 nixLog() {
     if [[ -z ${NIX_LOG_FD-} ]]; then
         return
     fi
-    echo "$@" >&"$NIX_LOG_FD"
+
+    local level="$1"
+    shift
+
+    if (( "${NIX_DEBUG:-0}" >= "$level" )); then
+        echo "$@" >&"$NIX_LOG_FD"
```

Comment:
```suggestion
        printf '%s\n' "$*" >&"$NIX_LOG_FD"
```

Nixkgs does this all over the place, but [`echo` is incredibly cursed](http://www.etalabs.net/sh_tricks.html), so if you’re touching this function we might as well fix `nixLog 2 -noooo!`.

This is definitely an internal interface that we can break whenever we want, right? Nothing else in‐tree uses it so hopefully nothing out‐of‐tree does either. But maybe it should fail loudly if it’s passed something that doesn’t look like a log level?

By the way `(( … ))` is also cursed; it will evaluate things like `NIX_DEBUG=1+1`. I recommend using `test` (aka the more confusing `[`).

Reply:
Thanks. I've updated the review based on this feedback.

Reply:
>By the way `(( … ))` is also cursed; it will evaluate things like `NIX_DEBUG=1+1`. I recommend using test (aka the more confusing `[`).

I think we should instead assert early on that `NIX_DEBUG` cannot be any other undocumented value (see https://nixos.org/manual/nixpkgs/stable/#var-stdenv-NIX_DEBUG), i.e.
```
[[ -z ${NIX_DEBUG-} || $NIX_DEBUG == [0-7] ]]
```

That is, `NIX_DEBUG` **must** be a documented integer level and we should be able to use it safely as an integer.

Reply:
I like that a lot, @tie.